src/middlewared/middlewared/plugins/auth_/authenticate.py

```diff
@@ -13,7 +15,7 @@ class Config:
     async def authenticate(self, username, password):
         if '@' in username:
```

Comment:
Since we have PAM library, you can just use that for all authentication other than root. You just need to have a special pam config file that you regenerate when directory services settings are changed (for middleware auth).

Reply:
Using PAM file will also allow us to avoid a datastore.query in potentially hot code path.

Reply:
Sample pam file for etc_files if you're not familiar (note didn't test this particular one):

```
#
# /etc/pam.d/middleware - authentication settings for middlewared service
#

<%namespace name="pam" file="pam.inc.mako" />\
<%
        dsp = None
        if middleware.call_sync('datastore.config', 'system.settings')['stg_ds_auth']:
            dsp = pam.getDirectoryServicePam(middleware=middleware, render_ctx=render_ctx)
%>\

% if dsp:
${dsp.pam_auth()}
% endif

auth    [success=1 default=ignore]      pam_unix.so obscure sha512
# here's the fallback if no module succeeds
auth    requisite                       pam_deny.so
# prime the stack with a positive return value if there isn't one already;
# this avoids us returning an error just because nothing sets a success code
# since the modules above will each just jump around
auth    required                        pam_permit.so
# and here are more per-package modules (the "Additional" block)
```

Reply:
In some cases we need to allow root user login even if it has no password set in `/etc/shadow`, is there a method to do this with pam?

Reply:
`root` would be the one exception to this. I can't think of a clever way to avoid this. Maybe if pam.authenticate() fails, we can have special error handling for root that will fallback to comparing the password hash with what's in our db. 

You'll probably want to specify `nodelay` for pam_unix.so (assuming we don't want 2+ second delays on failure path).

Reply:
In theory we could maintain a berkley db with local overrides and place pam_userdb.so ahead of pam_unix.so.